sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,163 @@
+            .. process response body
+           }
+      }
+
+      void OnError(nostd::string_view err) noexcept override
+      {
+          std::cout << " Error:" << err;
+      }
+  };
+
+  SessionManager sessionManager; // implementer can provide singleton implementation for it
+  auto session = sessionManager.createSession("localhost", 8000);
+  auto request = session->CreateRequest();
+  request->AddHeader(..);
+  SimpleResponseHandler res_handler;
+  session->SendRequest(res_handler);
+  session->FinishSession() // optionally in the end
+  ...shutdown
+  sessionManager.FinishAllSessions()
+*/
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class Method
+{
+  Get,
+  Post,
+  Put,
+  Options,
+  Head,
+  Patch,
+  Delete
+};
+
+enum class SessionState
+{
+  Created,             // session object is created
+  Ongoing,             // session is ongoing
+  Finished,            // session is finished ( this needs to be the final state )
+  Queued,              // http request is queued
+  TimedOut,            // Request timedout, no response received
+  Aborted,             // http request aborted due to local error,
+  Cancelled,           // http request cancelled, possibly due to session->CancelSession();
+  SendingFailed,       // http request sending failed
+  NetworkError,        // network error
+  SSLHandshakeFailed,  // ssl handshake failed
+  ReadError,           // error while reading response
+  WriteError           // error while writing rquest
+};
+
+using Byte           = uint8_t;
+using StatusCode     = uint16_t;
+using Body           = std::vector<Byte>;
+using SSLCertificate = std::vector<Byte>;
+
+class Request
+{
+public:
+  virtual void SetMethod(Method method) noexcept = 0;
+
+  virtual void SetUri(const nostd::string_view &uri) noexcept = 0;
+
+  virtual void SetBody(Body &body) noexcept = 0;
+
+  virtual void AddHeader(const nostd::string_view &name, const std::string &value) noexcept = 0;
+
+  virtual void ReplaceHeader(const nostd::string_view &name, const std::string &value) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+
+  virtual ~Request() = default;
+};
+
+class Response
+{
+public:
+  virtual const Body &GetBody() const noexcept = 0;
+
+  virtual bool ForEachHeader(
+      nostd::function_ref<bool(nostd::string_view name, std::string value)> callable) const
+      noexcept = 0;
+
+  virtual bool ForEachHeader(
+      const nostd::string_view &key,
+      nostd::function_ref<bool(nostd::string_view name, std::string value)> callable) const
+      noexcept = 0;
+
+  virtual StatusCode GetStatusCode() const noexcept = 0;
+
+  virtual ~Response() = default;
+};
+
+class EventHandler
+{
+public:
+  virtual void OnResponse(Response &) noexcept = 0;
+
+  virtual void OnError(SessionState, nostd::string_view &) noexcept = 0;
+
+  virtual void OnSslValidation(nostd::function_ref<SSLCertificate &()>) noexcept {}
```

Comment:
What is the exact reason for not having:
```c++
virtual void OnSslValidation(const SSLCertificate &) noexcept {}
```
Is it because the certificate might be expensive to obtain under certain circumstances?

Reply:
No, the idea was to be able to pass the SSL (pinned) public key from application to the `http_client` implementation, and let the implementation take care of validating it with peer certificate. But I think it would be much cleaner to instead pass the peer public key from implementation to application, and let  application do validation using pinned key/certificate. In that case, the method would look similar as you proposed. I will change it as below:

`virtual void OnConnecting(const SSLCertificate &) noexcept {})`

@maxgolov fyi

Reply:
Done